File: src/main/java/frc/chargers/utils/EpsilonEquals.java
package frc.chargers.utils;

public class EpsilonEquals {
    private static final double EPSILON = 1E-9;
    private EpsilonEquals() {}

    /**
     * Checks if 2 doubles are equal; correcting for floating point error.
     * Note: cannot be used as extension due to the double type being primitive.
     * Usage: <code>equivalent(2.0, 3.0 - 1.0)</code>
     */
    public static boolean equivalent(double a, double b) {
        return Math.abs(a - b) <= EPSILON;
    }
}

File: src/main/kotlin/frc/chargers/utils/EpsilonEquals.kt
@file:Suppress("unused")
package frc.chargers.utils

import com.batterystaple.kmeasure.dimensions.AnyDimension
import com.batterystaple.kmeasure.quantities.Quantity
import kotlin.math.abs

private const val EPSILON = 1E-9

/** A method that checks if 2 doubles are equal; correcting for floating-point error. */
infix fun Double.epsilonEquals(other: Double): Boolean =
    abs(this - other) < EPSILON

/** A method that checks if 2 quantities are equal; correcting for floating-point error. */
infix fun <D: AnyDimension> Quantity<D>.epsilonEquals(other: Quantity<D>): Boolean =
    siValue epsilonEquals other.siValue